src/utils/point_locator_nanoflann.C

```diff
@@ -0,0 +1,427 @@
+
+  return found_elem;
+}
+
+
+void
+PointLocatorNanoflann::operator() (const Point & p,
+                                   std::set<const Elem *> & candidate_elements,
+                                   const std::set<subdomain_id_type> * allowed_subdomains) const
+{
+  libmesh_assert (this->_initialized);
+
+  LOG_SCOPE("operator() returning set", "PointLocatorNanoflann");
+
+  // Make sure the output set is initially empty
+  candidate_elements.clear();
+
+  // Keep track of the number of elements checked in detail
+  unsigned int n_elems_checked = 0;
+
+  // Do the KD-Tree search
+  auto result_set = this->kd_tree_find_neighbors(p, _num_results);
+
+  for (auto r : make_range(result_set.size()))
+    {
+      // Translate the Nanoflann index, which is from [0..n_points),
+      // into the corresponding Elem id from the mesh.
+      auto nanoflann_index = _ret_index[r];
+      const Elem * candidate_elem = (*_elems)[nanoflann_index];
+
+      // Before we even check whether the candidate Elem actually
+      // contains the Point, we may need to check whether the
+      // candidate Elem is from an allowed subdomain.  If the
+      // candidate Elem is not from an allowed subdomain, we continue
+      // to the next one.
+      if (allowed_subdomains && !allowed_subdomains->count(candidate_elem->subdomain_id()))
+        continue;
+
+      // If we made it here, then the candidate Elem is from an
+      // allowed subdomain, so let's next check whether it contains
+      // the point. If the user set a custom tolerance, then we
+      // actually check close_to_point() rather than contains_point(),
+      // since this latter function warns about using non-default
+      // tolerances, but otherwise does the same test.
+      bool inside = _use_contains_point_tol ?
+        candidate_elem->close_to_point(p, _contains_point_tol) :
+        candidate_elem->contains_point(p);
+
+      // Increment the number of elements checked
+      n_elems_checked++;
+
+      // If the point is contained in/close to an Elem from an
+      // allowed subdomain, add it to the list.
+      if (inside)
+        candidate_elements.insert(candidate_elem);
+    } // end for(r)
+
+  // Debugging: for performance reasons, it may be useful to print the
+  // number of Elems actually checked during the search.
+  // libMesh::err << "Checked " << n_elems_checked << " nearby Elems before finding a containing Elem." << std::endl;
+}
+
+
+void
+PointLocatorNanoflann::enable_out_of_mesh_mode ()
+{
+  // Out-of-mesh mode should now work properly even on meshes with
+  // non-affine elements.
+  _out_of_mesh_mode = true;
+}
+
+
+void
+PointLocatorNanoflann::disable_out_of_mesh_mode ()
+{
+  _out_of_mesh_mode = false;
+}
+
+
+
+std::size_t
+PointLocatorNanoflann::get_num_results() const
+{
+  return _num_results;
+}
+
+
+
+void
+PointLocatorNanoflann::set_num_results(std::size_t val)
+{
+  // Must request at least 1 result
```

Comment:
Should we add an assert to that effect?

Reply:
We could I guess... but it seemed like such a pathological case that I was happy to just silently upgrade any such request from 0 to 1.